crates/symbolicator/src/sources.rs

```diff
@@ -125,38 +127,59 @@ pub struct FilesystemSourceConfig {
 }
 
 /// Local helper to deserialize an S3 region string in `S3SourceKey`.
-fn deserialize_region<'de, D>(deserializer: D) -> Result<rusoto_core::Region, D::Error>
+fn deserialize_region<'de, D>(deserializer: D) -> Result<Region, D::Error>
 where
     D: Deserializer<'de>,
 {
     // This is a Visitor that forwards string types to rusoto_core::Region's
     // `FromStr` impl and forwards tuples to rusoto_core::Region's `Deserialize`
     // impl.
-    struct RusotoRegion;
+    struct SdkRegion;
 
-    impl<'de> serde::de::Visitor<'de> for RusotoRegion {
-        type Value = rusoto_core::Region;
+    impl<'de> serde::de::Visitor<'de> for SdkRegion {
+        type Value = Region;
 
         fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
-            formatter.write_str("string or tuple")
+            formatter.write_str("string")
         }
 
-        fn visit_str<E>(self, value: &str) -> Result<rusoto_core::Region, E>
+        fn visit_str<E>(self, value: &str) -> Result<Region, E>
         where
             E: serde::de::Error,
         {
-            FromStr::from_str(value).map_err(|e| E::custom(format!("region: {:?}", e)))
-        }
-
-        fn visit_seq<S>(self, seq: S) -> Result<rusoto_core::Region, S::Error>
-        where
-            S: serde::de::SeqAccess<'de>,
-        {
-            Deserialize::deserialize(serde::de::value::SeqAccessDeserializer::new(seq))
+            // aws-sdk-rust does not validate region strings
+            const AWS_REGIONS: [&str; 21] = [
+                "ap-east-1",
+                "ap-northeast-1",
+                "ap-northeast-2",
+                "ap-south-1",
+                "ap-southeast-1",
+                "ap-southeast-2",
+                "ca-central-1",
+                "cn-north-1",
+                "cn-northwest-1",
+                "eu-central-1",
+                "eu-north-1",
+                "eu-west-1",
+                "eu-west-2",
+                "eu-west-3",
+                "sa-east-1",
+                "us-east-1",
+                "us-east-2",
+                "us-gov-east-1",
+                "us-gov-west-1",
+                "us-west-1",
+                "us-west-2",
+            ];
+            let pos = AWS_REGIONS.iter().position(|&x| x == value);
+            match pos {
+                Some(_) => Ok(Region::new(String::from(value))),
+                None => Err(E::custom(format!("unknown region: {}", value))),
```

Comment:
IIUC this means we need to updated our code every time they add a region?  That might not be ideal?

Reply:
I updated this code to use the EC2 `describe_regions` function and cache the result. Things I don't like about my solution:
* I'm calling `tokio::runtime::Runtime::new()` to grab a runtime so I can call `block_on` on it. This is needed because the `describe_regions` call uses an `async` `send` wrapper. Would it not be more appropriate to use one of the existing runtimes created in `server.rs`?
* I'm creating a `shared_config` using `aws_config::from_env().load().await`; not even sure this will work, given that it should really use the `credentials_provider` (`Container` or `Static`) associated with the current source. Thoughts?

Also, does the use of the `#[cached]` macro look OK?

Reply:
Ugh, this does indeed break the unit tests. Have to rethink this, because we need the region list during `serde` config parsing yet we won't have the credentials until _after_ parsing. And there's no way to get the region list anonymously beforehand so we can pre-populate the list.

Reply:
Yes, I'll do that.

By the way, Rusoto currently also hardcodes the list of [known regions](https://rusoto.github.io/rusoto/rusoto_core/region/enum.Region.html), so if a new region is added, a recompile is needed anyway once Rusoto is updated to include it.

Reply:
How about I add a simple test at startup to see if `sts get-caller-identity` works and if it fails, error out with an appropriate error message?